### ISSUE TITLE
Support building sdks that can build self-contained / ReadyToRun / NativeAOT applications without using pre-builts from nuget.org

### ISSUE THREAD

**Issue by tmds** [dotnet/source-build#1215]
Thanks to the reference packs, a source-build sdk can now build .NET Core applications that can be considered themselves source-buildable.

This doesn't apply to self-contained applications, and ReadyToRun applications. These require downloading a `Microsoft.NETCore.App.Runtime.<rid>` package from nuget.org. (This includes framework dependent ReadyToRun, probably because it uses the package to aquire `crossgen`.)

This issue is for building and including the Runtime package with the sdk. This makes it possible for the maintainer to apply patches during source-build, and the applications built this way using the sdk can be considered source-buildable.

cc @dagood @nguerrera @omajid

**Comment by dagood** [comment]
> This issue is for building and including the Runtime package with the sdk.

Runtime packs are similar to targeting packs, so IMO they should be included in e.g. `dotnet-runtime-pack-x-y` packages, which the SDK package would have a dependency on.

> This makes it possible for the maintainer to apply patches during source-build, and the applications built this way using the sdk can be considered source-buildable.

A note about this from https://github.com/dotnet/source-build/issues/1202, to put it in context: the runtime pack has limitations on what it supports, due to glibc compatibility and portable vs. non-portable builds. An SCD publish for `linux-x64` would need to bring down the extra-compatible prebuilt, but e.g. `fedora.30-x64` would be source-built.

---

I believe the SDK and perhaps Core-SDK repos will need to add support for this. The SDK [currently locates targeting packs with `TargetingPackRoot`](https://github.com/dotnet/sdk/blob/f16d813fa5671eea294073b76ef9019081034306/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L26) (`$(NetCoreTargetingPackRoot)`) which points to e.g. `/usr/bin/dotnet/packs/`. The [runtime pack resolver task](https://github.com/dotnet/sdk/blob/27f6210a6fca3b6b690c1b350d3ce9f46b930a70/src/Tasks/Microsoft.NET.Build.Tasks/ResolveRuntimePackAssets.cs) doesn't seem to have anything similar.

[The definition of `NetCoreTargetingPackRoot` is over in Core-SDK](https://github.com/dotnet/core-sdk/blob/dd98afbeb4d8d28856b8b2df8805d7dbd1614198/src/redist/targets/GenerateBundledVersions.targets#L156).

The [apphost pack resolver](https://github.com/dotnet/sdk/blob/27f6210a6fca3b6b690c1b350d3ce9f46b930a70/src/Tasks/Microsoft.NET.Build.Tasks/ResolveAppHosts.cs#L18) uses `TargetingPackRoot` too, despite the name, so I guess it might be fine to use it for runtime packs too.

**Comment by tmds** [comment]
It makes sense to make it work similar to the apphost package.

Can we set a target milestone for this?

I can probably implement this based on these references.

**Comment by nguerrera** [comment]
Adding @dsplaisted

**Comment by uweigand** [comment]
The current .NET 6 SDK already *checks* for runtime packs in the `packs/` subdirectory, it just doesn't find any because none were installed.  I've been able to do using the following straightforward patch to the installer:
```
Index: installer/src/redist/targets/GenerateLayout.targets
===================================================================
--- installer.orig/src/redist/targets/GenerateLayout.targets
+++ installer/src/redist/targets/GenerateLayout.targets
@@ -134,6 +134,13 @@
         <RelativeLayoutPath>packs/%(PackageName)/%(PackageVersion)</RelativeLayoutPath>
       </BundledLayoutPackage>

+      <BundledLayoutPackage Include="MicrosoftNetCoreAppRuntimePackNupkg">
+        <PackageName>Microsoft.NETCore.App.Runtime.$(SharedFrameworkRid)</PackageName>
+        <PackageVersion>$(MicrosoftNETCoreAppRefPackageVersion)</PackageVersion>
+        <TargetFramework>$(TargetFramework)</TargetFramework>
+        <RelativeLayoutPath>packs/%(PackageName)/%(PackageVersion)</RelativeLayoutPath>
+      </BundledLayoutPackage>
+
       <BundledLayoutPackage Include="MicrosoftNetStandardTargetingPackNupkg">
         <PackageName>NETStandard.Library.Ref</PackageName>
         <PackageVersion>$(NETStandardLibraryRefPackageVersion)</PackageVersion>
@@ -146,6 +153,13 @@
         <PackageVersion>$(MicrosoftAspNetCoreAppRefPackageVersion)</PackageVersion>
         <TargetFramework>$(TargetFramework)</TargetFramework>
         <RelativeLayoutPath>packs/%(PackageName)/%(PackageVersion)</RelativeLayoutPath>
+      </BundledLayoutPackage>
+
+      <BundledLayoutPackage Include="MicrosoftAspNetCoreAppRuntimePackNupkg">
+        <PackageName>Microsoft.AspNetCore.App.Runtime.$(SharedFrameworkRid)</PackageName>
+        <PackageVersion>$(MicrosoftAspNetCoreAppRefPackageVersion)</PackageVersion>
+        <TargetFramework>$(TargetFramework)</TargetFramework>
+        <RelativeLayoutPath>packs/%(PackageName)/%(PackageVersion)</RelativeLayoutPath>
       </BundledLayoutPackage>

       <BundledLayoutPackage Include="MicrosoftNetCoreAppHostPackNupkg">
```

This likely still needs a bit of tweaking to use an appropriate condition when to do it.   Also, it currently installs only the `$(SharedFrameworkRid)` ... maybe it also should install alternate RIDs as is done for the Apphost?

As a drawback, this approach significantly increases the size of the SDK installation on disk.  This is annoying in particular as most of the files installed as part of the target runtime pack are in fact *already present* in the SDK on-disk installation, under the `shared/` directory ... but it looks like that cannot be used as a drop-in replacement because the directory structure is different.

**Comment by tmds** [comment]
.NET 7 will add support for NativeAOT.

We'd like a source-build SDK to be able to build self-contained NativeAOT applications.

This issue was already tracking adding support for self-contained, and I've added NativeAOT in the title too. We can split it up when that makes sense.

**Comment by crummel** [comment]
I've summarized what I believe to be the current state of play and the agreed-upon solution for NativeAOT in 7.0 in https://github.com/dotnet/runtime/pull/76206.  Please check this out and let me know if you have any concerns.

**Comment by omajid** [comment]
@MichaelSimons you were asking about this issue earlier today?

**Comment by omajid** [comment]
@tmds would you agree this is more important than supporting building and using workloads from source?

**Comment by tmds** [comment]
Yes. These features apply to any app, while the workloads are for specific use-cases.

**Comment by tmds** [comment]
@richlander I'm wondering, do you think it would be beneficial in the future if Microsoft sdk images include some of these features (so: without having to reach out to nuget.org) as the features are becoming more prevalent?

**Comment by ashnaga** [comment]
For self-contained deployment -

Self-contained deployment acquires runtime packs and AppHost packs (cross-target) from NuGet (MS Built). These are not packaged with SDK due to the size constraint. And so, as the packages are not offered with Linux distribution-built SDK as well and end-up acquired from the NuGet. To acquire packages which are source-built by Linux distributions, here are few options.
-	Package with SDK (if small size). Starting .NET 6, SDK does check for runtime packs and AppHost packs in dotnet/packs folder first before acquiring from NuGet. Alternatively, when user opt to use Linux package manager to get those packages, by default Install to dotnet/packs folder.
-	Another option is to build and offer those as NuGet package, but I’m not sure if maintainers will prefer this option. 

A few points to note –
-	Linux distributions only build packages specific to their distribution. As in RHEL only builds runtime pack rhel-x64 and not linux-x64 or ubuntu-x64. So, the size of packages to offer must be smaller compared to Microsoft Build.
-	Source-built does offer Linux distros to build AppHost and Runtime packages.
-	For cross-targeting scenarios, ex- user building on rhel-x64 and target platform being ubuntu-x64. This would still require pulling packages from NuGet.org (MS Built) as Linux distributions do not build packages for other distributions. So, if packages are not acquired with distributions preferred mechanism, then the packages will be acquired using NuGet.org.

**Comment by uweigand** [comment]
> These are not packaged with SDK due to the size constraint.

I was wondering about just that size issue.  The main problem appears to me that if we place a copy of the `Microsoft.NETCore.App.Runtime.<rid>` package under the `packs/` directory, then all those files are present twice: under `packs` *and* under `shared/Microsoft.NETCore.app`.

Is this duplication really necessary?   Would it be possible to rework the dotnet loader so it takes all the runtime assemblies (and other files) from the `packs/` directory if present, so we would then just not need the `shared/` directory at all?  Or am I missing something here?

**Comment by tmds** [comment]
I also think size is an important factor, esp if we'd like to include these features in a container image.

The duplication isn't needed, but it comes from staying close to how things work now with the nuget.org packages.
And since that is working well, I imagine there isn't much interest upstream in exploring the design to minimize the size. And besides the effort, there's also the risk of breaking things by entangling them.

> For self-contained deployment

What about R2R and NativeAOT?

**Comment by dsplaisted** [comment]
The idea has come up before that we could use the shared framework from `shared` instead of the runtime pack from `packs` in order to save space or downloads.  I think that even though they are mostly the same, there are still some differences.  It's possible those differences wouldn't apply to source-built versions of the .NET SDK.  I think @jkotas knows the details.

**Comment by jkotas** [comment]
We do not want to have hard coupling between the runtime used by the build output with the runtime that that the SDK happens to run on. The output of the build can use different (servicing or major) version, different architecture, different runtime flavor (mono, aot, ...), etc.

In theory, we can special case the case where the shared runtime exactly matches the runtime that the self-contained deployment wants to use, and skip the download of the runtime pack in that case. We did not think that the self-contained publish to be common enough for this special casing and all complications that come with it to be worth it.

**Comment by jkotas** [comment]
> I also think size is an important factor, esp if we'd like to include these features in a container image.

This size factor can be fixed by linking the files to the same physical file.

**Comment by tmds** [comment]
> This size factor can be fixed by linking the files to the same physical file.

I was wondering about that.

Whether it could be symbolic links, or have to be hard links?
And if the runtime repo could provide such an output, or whether source-build repo could/should do it?

**Comment by jkotas** [comment]
> Whether it could be symbolic links, or have to be hard links?

I do not know. Some tools may have problems with symbolic links.

> And if the runtime repo could provide such an output, or whether source-build repo could/should do it?

Build of any more complicated app ends up restoring something from nuget. I do not think we want to include more stuff in the default SDK or otherwise complicate things to fine-tune the amount of stuff that gets downloaded. For example, we have long-standing issue in that we download all runtime packs (even ASP.NET Core and Windows Deskop packs) when publishing self-contained console apps that we did not care enough to fix.

If the distro builds want to include more packages in place (and potentially optimize the amount of diskspace used by linking), I think it would be a specific to just the distro builds.

**Comment by tmds** [comment]
> Some tools may have problems with symbolic links.

Yes, symbolic links may not work for different reasons.

And, hard links come with a packaging challenge because the hard links link to files that are in a different package.
For example, a pack package with a hard link to a file that is provided by a runtime package.

Though the archive format may support this, the default archive tooling doesn't*.
So packages with hard links need to be crafted.

(*: it can't distinguish between what package provides, and what package hard links)

**Comment by uweigand** [comment]
> And, hard links come with a packaging challenge because the hard links link to files that are in a different package. For example, a pack package with a hard link to a file that is provided by a runtime package.

Couldn't the runtime package just automatically always provide both the shared runtime and the runtime pack?  If this is hard links, there wouldn't be any disk space reason not to, right?

**Comment by tmds** [comment]
> Couldn't the runtime package just automatically always provide both the shared runtime and the runtime pack? If this is hard links, there wouldn't be any disk space reason not to, right?

Yes, though the runtime pack is used by the sdk, we could ship it in the runtime package (with no additional size cost).

We also need to package the R2R and NativeAOT tooling.

**Comment by tmds** [comment]
@ashnaga @MichaelSimons can we for an upcoming preview of .NET 8 target to include the runtime pack as part of the vmr output?

Once we have that, we can then for the subsequent preview focus on reducing its size either by using symbolic or hard links.

And, can you propose how we could include R2R and NativeAOT?

**Comment by jkotas** [comment]
> can we for an upcoming preview of .NET 8 target to include the runtime pack as part of the vmr output?

I would expect that vmr is building the runtime pack and R2R .nupkg files. Is that not the case?

**Comment by tmds** [comment]
> I would expect that vmr is building the runtime pack and R2R .nupkg files. Is that not the case?

Yes, some of these nuget packages get built.

The output of the vmr is "DOTNET_ROOT" file layout which is then packaged up by the distro maintainers.
The request include the files from the nuget package at the appropriate location.

**Comment by tmds** [comment]
> @ashnaga @MichaelSimons can we for an upcoming preview of .NET 8 target to include the runtime pack as part of the vmr output?

Can we do this?

**Comment by tmds** [comment]
@ashnaga @MichaelSimons what's the next step?

**Comment by MichaelSimons** [comment]
@ashnaga is driving this one.  I will let her respond.

**Comment by ashnaga** [comment]
@tmds 
Thank you for your patience. I made an evaluation to ensure that packages are picked from dotnet/packs folder if available and not being pulled from NuGet. 
Currently, if there is a package with following naming convention - "dotnet/packs/Microsoft.NETCore.App.Runtime.linux-x64" then only it will be picked by SDK. The naming convention does not imply distribution/RID specific packages which we are using in case of AppHost packages - https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging

In order to make it work for distribution/RID specific packages, the source-built version will need to change how the Bundled versions targets file is generated:
https://github.com/dotnet/installer/blob/7eea2f60fdcf419d65872c6173f2bdcd8381905a/src/redist/targets/GenerateBundledVersions.targets#L196-L212

@tmds 
To make this work in preview timeframe, are you happen to make these changes?

cc @dsplaisted @MichaelSimons if you have specific questions on changes in the mentioned file.

**Comment by tmds** [comment]
> To make this work in preview timeframe, are you happen to make these changes?

If no one else is available that is more familiar, I will look into it.

In the meanwhile we should try to determine what is needed to provide NativeAOT. Then we can start working on that too.

**Comment by tmds** [comment]
@ashnaga do you have a picture of what is needed for NativeAOT?

A heads up: I'll be on summer leave from mid July to mid August. It would be good if we can make some more progress before that.

**Comment by jkotas** [comment]
> do you have a picture of what is needed for NativeAOT?

- Add dotnet/llvm-project to https://github.com/dotnet/dotnet, and make it build objwriter there. (It should be built before dotnet/runtime.)
- Find and delete all places where we are disabling native AOT for source build. For example: https://github.com/dotnet/runtime/blob/aca0eab8fcc5e4c82f7895b017595788752dc3f7/src/coreclr/tools/aot/crossgen2/crossgen2.csproj#L7-L8
- Bug-fix level work to make things actually work.
- Figure out how to distribute the packages (discussion above)

**Comment by tmds** [comment]
> Add dotnet/llvm-project to https://github.com/dotnet/dotnet, and make it build objwriter there. (It should be built before dotnet/runtime.)

This requires adding some things to installer `src/SourceBuild/content/repo-projects`.
With the move to the vmr, I don't know how to go about building that.

@MichaelSimons can you provide some help?

**Comment by premun** [comment]
@tmds this should be a starting point for adding new repositories into Source Build: https://github.com/dotnet/source-build/blob/main/Documentation/sourcebuild-in-repos/new-repo.md

**Comment by tmds** [comment]
> this should be a starting point for adding new repositories into Source Build: https://github.com/dotnet/source-build/blob/main/Documentation/sourcebuild-in-repos/new-repo.md

I'm looking for the command that you can run on the `installer` repo to make it build up the complete sources (which is what is put into `dotnet/dotnet`). I don't see it mentioned in these docs.

**Comment by premun** [comment]
I think the right way is to clone the VMR, manually put the `llvm` in `src/`. Register the repository (`repo-project`...) and build it.

Once you get that working, we will open a PR against `installer` to put the `repo-project` file under `src/VirtualMonoRepo` and also change the [`source-mappings.json`](https://github.com/dotnet/installer/blob/main/src/VirtualMonoRepo/source-mappings.json). Then, when you open a PR, the PR build will validate everything.

You can also start with a PR and work your way through that but it will possibly cost a lot of iterations and compute as the PR build builds much more than the VMR.

**Comment by premun** [comment]
Couple of notes to the above: To make everything work in the PR build of installer, we will also have to make sure the LLVM flows into the VMR which means it would have to be declared as a dependency in one of the repositories (in `eng/Version.Details.xml`). It is still possible to have the following workflow:

1. Make a dev branch in a fork of `installer`
2. Declare the dependency on `llvm-project` there (in `eng/Version.Details.xml`)
3. Add the `llvm-project` repo to `src/VirtualMonoRepo/source-mappings.json` in the `installer` dev branch
4. Push the dev branch into your fork and start a Codespace (you have to select the VMR `.devcontainer`)

The installer VMR-specific Codespace is configured to create and sync the VMR into `/workspaces/dotnet` so it should bring in the `llvm-project` there.

In this workflow, you can:
- Keep pushing new changes into the `installer` dev branch and awlays start a fresh Codespace
- In the `installer` branch, you can already add the new `repo-project` files and everything - this way your changes will persist between attempts. 
- Build the VMR in the Codespace right away to test things as you go

Do not open an installer PR in this workflow - it's not required. Also I am not sure if you will need to have billing set up for the Codespace to work for you.

**Comment by MichaelSimons** [comment]
> Something like the following, but what to use as the Name and Version?

See https://github.com/dotnet/installer/pull/14697 specifically https://github.com/dotnet/installer/pull/14697/files#diff-fb62e94a1d6f29f863e3d0a22aa38269f6cd1d7f03b109dc06e2cbf2548b86d3L20.  This shouldn't use the SourceBuildTarball element as we are trying to remove it.  Instead it should use SourceBuild but that requires LLVM be onboarded to source-build so that it has a source-build leg and publishes a source-build intermediate.

**Comment by tmds** [comment]
I've opened two PRs for getting llvm-project in the vmr: https://github.com/dotnet/installer/pull/16757, https://github.com/dotnet/llvm-project/pull/439.

**Comment by tmds** [comment]
@ashnaga @MichaelSimons @premun 

Based on where we are with source-buildable NativeAOT, it makes sense to defer this to .NET 9 so we can focus our efforts on the upcoming .NET 8 release.

**Comment by ashnaga** [comment]
@jkotas @MichalStrehovsky 
Does this merge of [ObjWriter in C#](https://github.com/dotnet/runtime/issues/77178#issuecomment-1816106564) impact the instruction on what is needed to enable NativeAOT with source-build [instructions](https://github.com/dotnet/source-build/issues/1215#issuecomment-1569490565).

[PR](https://github.com/dotnet/installer/pull/16757#issuecomment-1923825759)to build  llvm-project is closed now, considering this as an [alternative](https://github.com/dotnet/runtime/issues/77178#issuecomment-1815901537).

**Comment by tmds** [comment]
We also need to some work so the ILCompiler pack becomes part of the source-built SDK that is built by the vmr.

The work should be similar to the runtime pack we've added in .NET 8 to enable self-contained publish.

I'll look into that after looking at the bug-fix.

**Comment by tmds** [comment]
I think the SDK needs these 3 NuGet packages to publish a NativeAOT application:

- Microsoft.DotNet.ILCompiler
- Microsoft.NET.ILLink.Tasks
- runtime.<rid>.Microsoft.DotNet.ILCompiler

Like with self-contained, we'd like to provide NativeAOT through distro packages.

For self-contained we placed the files from `Microsoft.NETCore.App.Runtime.<rid>` and `Microsoft.AspNetCore.App.Runtime.<rid>` under `DOTNET_ROOT/packs`.

I think the SDK only looks at `DOTNET_ROOT/packs` for specific types of packages, so we can't use the same mechanism for these 3 NuGet packages, can we?

Is there some other mechanism we can use?

cc @dsplaisted @jkotas

**Comment by dsplaisted** [comment]
> I think the SDK only looks at `DOTNET_ROOT/packs` for specific types of packages, so we can't use the same mechanism for these 3 NuGet packages, can we?
> 
> Is there some other mechanism we can use?
> 
> cc @dsplaisted @jkotas

I think we would need changes in the .NET SDK to enable this.  I think we should put these assets in the packs folder in the same way as the other assets, and modify the SDK to pick them up from there.

Basically use the same mechanism that we already use for other packs, but add support for using it for these ones.

**Comment by tmds** [comment]
I've opened 2 PRs: https://github.com/dotnet/runtime/pull/99148 and https://github.com/dotnet/installer/pull/18835.

With these PRs, the resulting SDK can be used with the 3 source-built NuGet packages (`Microsoft.DotNet.ILCompiler`, `Microsoft.NET.ILLink.Tasks`, `runtime.<rid>.Microsoft.DotNet.ILCompiler`) to publish NativeAOT apps!

I also did a second stage build and validated it works.

> I think we would need changes in the .NET SDK to enable this. I think we should put these assets in the packs folder in the same way as the other assets, and modify the SDK to pick them up from there.

I'm not sure how this will look. Will the NuGet packages that are available through packs be defined in `Microsoft.NETCoreSdk.BundledVersions.props`? Should we create an issue for adding this feature to the SDK?

**Comment by tmds** [comment]
> https://github.com/dotnet/runtime/pull/99148 and https://github.com/dotnet/installer/pull/18835.

These have been merged.

The next step is to work out how we deliver this in rpms (distro packages).

@dsplaisted is this something you can do? Or describe how it should work, so I can help implement it?

**Comment by ashnaga** [comment]
@dsplaisted @jkotas @baronfel
In case of Microsoft SDK, are these packages packaged with SDK or being acquired from NuGet during NativeAOT operations like build/publish?

Microsoft.DotNet.ILCompiler
Microsoft.NET.ILLink.Tasks
runtime.<rid>.Microsoft.DotNet.ILCompiler

**Comment by jkotas** [comment]
They are acquired from NuGet for Microsoft SDK. We do not want to include them in the default Microsoft SDK install due to size.

**Comment by baronfel** [comment]
I was doing a bit of investigation into the [`ProcessFrameworkReferences`](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs) MSBuild Task, which is where most of the 'should we download anything from NuGet' logic lives for the runtime packs and so on, and it doesn't appear like we even support loading `Microsoft.DotNet.ILCompiler`, `Microsoft.NET.ILLink.Tasks`, and `runtime.**RID**.Microsoft.DotNet.ILCompiler` from the `packs` directory at all - they are always assumed to be acquired from NuGet. So I expect that for a source-built SDK to light up in the presence of these packs we'd need to change `ProcessFrameworkReferences` to support [locating Tool Packs](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L720) from the [`TargetingPackRoot`](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L899-L937) like we do for [runtime packs](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L662) and [targeting packs](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L255)

**Comment by tmds** [comment]
> So I expect that for a source-built SDK to light up in the presence of these packs we'd need to change ProcessFrameworkReferences to support [locating Tool Packs](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L720) from the [TargetingPackRoot](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L899-L937) like we do for [runtime packs](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L662) and [targeting packs](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L255)

Do we need a new `ToolPackType`? which? And would, like for the other packs, the packages be listed under `Microsoft.NETCoreSdk.BundledVersions.props`?

**Comment by baronfel** [comment]
The design would be something like 
> Known Package References can either be located from the Workloads installed in a given SDK, the SDK's TargetingPackRoot, or from the configured user NuGet feeds. When locating a pack of any type, first the workloads installation pack directory will be probed for that pack, then the TargetingPackRoot location will be probed for that pack. If neither of those locations can satisfy the pack, the pack will be acquired as an implicit PackageReference

Practically speaking, this means that the method that currently unconditionally creates an implicit PackageReference for the ToolPacks would instead follow the pattern that runtime packs and targeting packs use in their matching methods:
* [Probe the pack paths for the pack](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L255)
* [if found, set metadata that points to the statically-known location](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L256-L261)
* [If not found, and the current build environment supports package downloads, then create an implicit package reference for the pack](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L264-L274)

**Comment by tmds** [comment]
@baronfel can you also answer my questions here: https://github.com/dotnet/source-build/issues/1215#issuecomment-2020305034?

**Comment by baronfel** [comment]
I don't think we'd need new toolpack types - the existing toolpacks just would all become viable targets for this kind of sdk-local location? @dsplaisted anything you see wrong with that? There's kind of a philosophical difference here where the MS-built SDKs wouldn't use this because we'd rather float to package feeds, but the source-built SDKs want to be more batteries-included, and this is our existing pattern for making batteries-included work. 

cc @jkotas do you have any thoughts on this?

**Comment by jkotas** [comment]
> cc @jkotas do you have any thoughts on this?

I do not have anything to add.

**Comment by tmds** [comment]
> [Probe the pack paths for the pack](https://github.com/dotnet/sdk/blob/5d0c38bf4aeebd7e28ae178c6e97dda8878018b9/src/Tasks/Microsoft.NET.Build.Tasks/ProcessFrameworkReferences.cs#L255)

How does `ProcessFrameworkReferences` know what packages it should probe the `packs` directory for, and when?

**Comment by dsplaisted** [comment]
> How does `ProcessFrameworkReferences` know what packages it should probe the `packs` directory for, and when?

It's in the logic of `ProcessFrameworkReferences`-- when it gets to a pack it needs, it calls `GetPackPath` to see if it's already available in various pack folders, and if so resolves it there.  If not, then it adds a PackageDownload to download the pack.

Over time, the set of packs that the task would do this for has grown, as we had more scenarios for including different types of packs in the SDK install.

The problem with the ILCompiler and ILLink packs is that they are referenced as `PackageReference` instead of `PackageDownload` items and have MSBuild `.props` and `.targets` files in them that need to be imported.  We don't have a good way of importing targets files from something resolved to the packs folder by the `ProcessFrameworkReferences` task, as by the time the task runs evaluation has already happened and we can't then import an additional .targets file.

Some ideas for what we could do:

- Move the .targets files for ILLink and ILCompiler from the NuGet packages to the .NET SDK.  The NuGet package would then be used just for delivering the tool binaries, but not the targets that call them.  I'm not sure how tightly coupled the targets are with the tools and how hard it would be to extract the targets separately.
- Include the .nupkg file itself in the `library-packs` folder of the .NET SDK.  NuGet would restore it from there and extract it to the packages folder.  We might want to use a different package identity for the source-built versions of these packages so there's no chance that the Microsoft-built ones from NuGet would get restored instead.

I think the first option is overall better, but the second one may be easier to implement.

**Comment by jkotas** [comment]
> Move the .targets files for ILLink and ILCompiler from the NuGet packages to the .NET SDK. The NuGet package would then be used just for delivering the tool binaries, but not the targets that call them. I'm not sure how tightly coupled the targets are with the tools and how hard it would be to extract the targets separately

These targets are tightly coupled with the compiler binary. Doing this would effectively break the ability to override the NativeAOT package version to workaround bugs, test privates, etc.

**Comment by tmds** [comment]
> We might want to use a different package identity for the source-built versions of these packages so there's no chance that the Microsoft-built ones from NuGet would get restored instead.

Also for the user, I think we should try to only use the source-built versions when the user explicitly opts in by specifying the source-built rid.

> The problem with the ILCompiler and ILLink packs is that they are referenced as PackageReference instead of PackageDownload items and have MSBuild .props and .targets files in them that need to be imported. We don't have a good way of importing targets files from something resolved to the packs folder by the ProcessFrameworkReferences task, as by the time the task runs evaluation has already happened and we can't then import an additional .targets file.

Could we extend `PackageReference` so `ProcessFrameworkReferences` can output PackageReferences that are found as extracted packages in the `packs` folder?

```
        PackageReference=
            Microsoft.DotNet.ILCompiler
                    IsImplicitlyDefined=true
                    PrivateAssets=all
                    PackageDirectory=/usr/lib64/dotnet/packs/Microsoft.DotNet.ILCompiler/8.0.3
                    Version=8.0.3
...
```

**Comment by tmds** [comment]
> Could we extend PackageReference so ProcessFrameworkReferences can output PackageReferences that are found as extracted packages in the packs folder

This suggestion is meant to meet the needs of making `ProcessFrameworkReferences` capable of using the source-built assets when needed, and continue use their `.props` and `.targets` as it is working now.

@dsplaisted @jkotas @baronfel what do you think?

**Comment by dsplaisted** [comment]
> > Could we extend PackageReference so ProcessFrameworkReferences can output PackageReferences that are found as extracted packages in the packs folder
> 
> This suggestion is meant to meet the needs of making `ProcessFrameworkReferences` capable of using the source-built assets when needed, and continue use their `.props` and `.targets` as it is working now.
> 
> @dsplaisted @jkotas @baronfel what do you think?

No, NuGet doesn't support setting a PackageDirectory on a PackageReference where the package is already extracted.

It sounds like the best solution is probably to put the source-built ILCompiler and ILLink NuGet packages in the `library-packs` folder.

**Comment by tmds** [comment]
> It sounds like the best solution is probably to put the source-built ILCompiler and ILLink NuGet packages in the library-packs folder (as .nupkg files).

There are some downsides to this:

The packages that are produced by source-build don't have stripped symbols which bloats them significantly. This is due to the stripping being performed at the end of the packaging process to produce separate debug packages.

In general (I don't know how it applies to these packages specifically) the source-build packages can be different from the Microsoft built packages due to limitations in what can be source-built. This can cause functional differences.

Afaik, NuGet doesn't have a preference between different feeds. We can't force it to prefer `library-packs` over nuget.org packages.

With the files held in a nupkg archive, we can't try to optimize the size by referencing common files like we have done when adding support for self-contained publish.

> No, NuGet doesn't support setting a PackageDirectory on a PackageReference where the package is already extracted.

Is it a feature we can consider?

**Comment by baronfel** [comment]
@dsplaisted and I chatted about this just now and we see two potential solutions, with short- and long- time frames:
 
### Short term
 
We think the least-friction, least-dependency way to get this working is to use library-packs. This would imply source-build producing the packages above (cc @jkotas - is there any set of functionality that source-built versions of these packages would lack compared to the MS-produced nuget.org versions?), and changing the names of the two without runtime-specific package identifiers. This would allow us to not have to worry about content mismatches if/when the MS-produced packages would be used. You would need to handle the symbol concerns you mention above, but the up-side to this is that you would be entirely in control - no need to wait on any other teams inside Microsoft to unblock this functionality.
 
### Longer term
 
Longer term, we'd want to work with the IL Compiler team to split out the contents of the above packages into completely separate components for build logic vs tools/binaries, ensuring that the tools are more standalone. With this done, they could be treated as Tool Packs in the `ProcessFrameworkReferences` Task and just laid down into an expected place in the SDK for the build logic to assume.

**Comment by jkotas** [comment]
> is there any set of functionality that source-built versions of these packages would lack compared to the MS-produced nuget.org versions?

They should be identical.

> changing the names of the two without runtime-specific package identifiers

The names of these packages are hardcoded in number of other places. Would this mean that all those other places have to have some sort of condition for including the package under a new name?

**Comment by jkotas** [comment]
@tmds How do you envision regular package references working in source-buildable apps? What is going to guarantee that the source-building of the apps uses the copy of the package built from source and not the copy of the package on nuget.org?

**Comment by dsplaisted** [comment]
> The names of these packages are hardcoded in number of other places. Would this mean that all those other places have to have some sort of condition for including the package under a new name?

More or less.  It could be a condition evaluated when the SDK is being built, or a condition in the SDK evaluated when building an app, or it could be a source-build patch that changes the name.

> What is going to guarantee that the source-building of the apps uses the copy of the package built from source and not the copy of the package on nuget.org?

I think the package identity should be different, for example `Redhat.DotNet.ILCompiler`, and the only place that package would be built would be via source-build.

**Comment by jkotas** [comment]
> I think the package identity should be different, for example Redhat.DotNet.ILCompiler

I meant how does the source-build work for random 3rd party packages. Let's say I want to source-build an app that references https://www.nuget.org/packages/Serilog package. Where is the source-built version of Serilog package going to show up in the system and how is it going to be referenced by the app?

I would like to see whether it would make sense to naturally extend the solution for 3rd party packages to packages like ILCompiler.

**Comment by tmds** [comment]
> I meant how does the source-build work for random 3rd party packages. Let's say I want to source-build an app that references https://www.nuget.org/packages/Serilog package. Where is the source-built version of Serilog package going to show up in the system and how is it going to be referenced by the app?

There is an open issue for that: https://github.com/dotnet/core/issues/7443. It might involve using nuget packages from a local folder that are all source-built and having no other feeds available, like when source-building .NET.

> I would like to see whether it would make sense to naturally extend the solution for 3rd party packages to packages like ILCompiler.

I think ILCompiler is different because the integration/coupling with the SDK.

**Comment by jkotas** [comment]
> I think ILCompiler is different because the integration/coupling with the SDK

I would not call it coupling. We have been intentionally keeping the AOT compiler package very loosely coupled with the SDK. The integration is basically just about automatically auto-referencing the ILCompiler package when the PublishAot property is set to provide better UX. We lived without this integration in the early days of the AOT project and it worked just fine for years. (The projects were required to reference the ILCompiler package manually.)

I think it would be best to treat the ILCompiler nuget package just like any other nuget package that the app source building may need. 

> It might involve using nuget packages from a local folder that are all source-built and having no other feeds available, like when source-building .NET.

This makes sense to me.

**Comment by jkotas** [comment]
Package source mapping https://learn.microsoft.com/en-us/nuget/consume-packages/package-source-mapping#enable-by-manually-editing-nugetconfig can help with mixing and matching local packages and nuget.org packages.

**Comment by tmds** [comment]
Where is the `library-packs` NuGet directory added as a NuGet source? Is it something the SDK does unconditionally?

> NuGet doesn't support setting a PackageDirectory on a PackageReference where the package is already extracted.

Why aren't we considering implementing this option?

My understanding is that it would allow to make `ProcessFrameworkReferences` to pick the source-built ILCompiler (and friends) from `packs` when needed.

**Comment by dsplaisted** [comment]
> Renaming packages in order to use them in source build feels wrong to me.

My opinion is that this is OK.  These packs or packages are implicitly referenced by the SDK.  If they were package references in a project file and you had to update that identity in order to use source-built packages, that would not be so nice.  It sounds like maybe they can be referenced separately, but the main scenario today is when they're automatically referenced by the SDK.

> I think it would be best to treat the ILCompiler nuget package just like any other nuget package that the app source building may need.

I don't think we know how we would handle source-built NuGet packages, and it could well be a large feature.  If we want to support source-built NativeAOT soon, then it's probably better not to wait for general source-built NuGet package support.

> Where is the `library-packs` NuGet directory added as a NuGet source? Is it something the SDK does unconditionally?

Yes, it does it unconditionally in [Microsoft.NET.NuGetOfflineCache.targets](https://github.com/dotnet/sdk/blob/544299b1466f508b041c7355f24b3215b4628991/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.NuGetOfflineCache.targets).  The `library-packs` folder was added to support workloads, but it is included as a source unconditionally and would work here.

> > NuGet doesn't support setting a PackageDirectory on a PackageReference where the package is already extracted.
> 
> Why aren't we considering implementing this option?

It seems to me that this would conflict with some design aspects of NuGet.  When NuGet extracts packages, it computes hashes of the packages and may verify that the package was correctly signed.  This information is saved to the extracted package folder and helps verify and maintain the integrity of the package.  If you try to point it to a folder you've created where the package is already extracted, then it won't have that extra information and I think NuGet won't be able to maintain the guarantees it tries to make.

I'm not an expert on a lot of the details of NuGet, so I could be wrong.  But even then, it would take time to design and implement a new NuGet feature for this.  If possible, I would recommend trying to go with a design that can be implemented mostly with existing functionality of the SDK and its components.

**Comment by jkotas** [comment]
> If we want to support source-built NativeAOT soon, then it's probably better not to wait for general source-built NuGet package support.

Creating steps for how one can use source-built Native AOT makes sense to me (e.g. using a custom Package Source Mapping rules that I have linked above). However, I do not think it make sense to spend a lot of one-off effort on polishing the experience for source-built NativeAOT without solving the larger problem of source-built nuget packages.

**Comment by tmds** [comment]
> I do not think it make sense to spend a lot of one-off effort on polishing the experience for source-built NativeAOT without solving the larger problem of source-built nuget packages.

They are related, but I don't expect https://github.com/dotnet/core/issues/7443 to become a high priority for .NET upstream in the near future.

For this issue we focus on a _.NET SDK distro maintainer_ providing an SDK that includes NativeAOT support as a built from source feature.

For https://github.com/dotnet/core/issues/7443 we focus on a _.NET app distro maintainer_ building an app from source.

These are different people.

For the second, some of the app maintainers already go through the effort of dealing with building their dependencies from source, and https://github.com/dotnet/core/issues/7443 is about providing tooling to assist them.

The second issue is mostly about dealing with 3rd party NuGet packages. I don't consider the ILCompiler a 3rd party NuGet package. I think we should try to deliver the source-built NativeAOT (as .NET SDK distro maintainers) so it just works for the SDK user (.NET app distro maintainer or regular distro user). They shouldn't be aware/care how the ILCompiler gets packaged with the SDK.

> I'm not an expert on a lot of the details of NuGet, so I could be wrong. But even then, it would take time to design and implement a new NuGet feature for this. If possible, I would recommend trying to go with a design that can be implemented mostly with existing functionality of the SDK and its components.

I'm not an expert either but it seems like an option that would give us much flexibility.

> The packages that are produced by source-build don't have stripped symbols which bloats them significantly. This is due to the stripping being performed at the end of the distro packaging process to produce separate distro debug packages.

Some numbers: the Microsoft ILCompiler nupkg is 48MB and the source-built is 126MB. Uncompressed they are 142MB and 326MB. I think we should try to strip those symbols to reduce the package size. I'm not sure where to strip them as we'd still want the non-stripped files to flow through for distro packaging. If the files are stripped inside a nupkg, the distro won't know about them, and it can't provide debug symbols through debug packages either.

**Comment by jkotas** [comment]
> For this issue we focus on a .NET SDK distro maintainer providing an SDK that includes NativeAOT support as a built from source feature.

It would help to describe the user scenario that this about. Why do the users want to use this and what do they expect to get out of it? My assumption has been that the motivating scenario for this work is ".NET app distro maintainer building an app from source", but it sounds like that it is not the case.

> The second issue is mostly about dealing with 3rd party NuGet packages. I don't consider the ILCompiler a 3rd party NuGet package.

Yes, native AOT toolchain is a .NET feature delivered via NuGet packages. We have many other .NET Runtime and SDK features that are delivered via NuGet packages. My point is that we should keep that in mind and design something that can work for any .NET feature delivered via NuGet package. The `library-packs` idea is going in this direction that's good.

(I would expect the solution to be naturally extendable to 3rd party packages as well, but I agree that building 3rd packages from source is a separate problem.)

**Comment by tmds** [comment]
> It would help to describe the user scenario that this about. Why do the users want to use this and what do they expect to get out of it? My assumption has been that the motivating scenario for this work is ".NET app distro maintainer building an app from source", but it sounds like that it is not the case.

The reason is the same as why .NET is buildable by others: it enables 3rd parties to build/deliver/support .NET independent of Microsoft.

The app maintainer is an important use-case.

The reason the app maintainer isn't allowed to use the NativeAOT because the distro want to be capable of changing/fixing the feature.

This applies also to RHEL: Red Hat won't claim we support NativeAOT on RHEL if we can't build it from source. The best we can do (and say) is: we work with Microsoft to ensure NativeAOT works well on RHEL.

Another use-case is to be able to build the feature on a distro/architecture that is not supported by Microsoft.

> Yes, native AOT toolchain is a .NET feature delivered via NuGet packages. We have many other .NET Runtime and SDK features that are delivered via NuGet packages. My point is that we should keep that in mind and design something that can work for any .NET feature delivered via NuGet package. The library-packs idea is going in this direction that's good.

I listed what I think are the main challenges with the nuget packages in source-build in an earlier comment: https://github.com/dotnet/source-build/issues/1215#issuecomment-2090941658.

**Comment by jkotas** [comment]
> I listed what I think are the main challenges with the nuget packages in source-build in an earlier comment: #1215 (comment).

Well, the alternative is to create a parallel path through the whole system where the native AOT toolchain does not look like a package. It sounds very complicated to me. I think addressing the issues with the package-centric approach is easier.

**Comment by dsplaisted** [comment]
Hi folks,

Some updates, and (I think) some good news.

While thinking this over I eventually came up with an idea that I think would let .props and .targets files be imported from packs in the `packs` folder.  The idea is that `ProcessFrameworkReferences` would output the list of files to be imported, and the SDK would have a target that ran during restore that would write out generated .props and .targets files that would import the specified files and would be automatically imported during the build.  This is how NuGet supports importing .props and .targets files from packages, so the SDK would do something similar also during restore.

However, while I was writing up a design for this idea, I looked more at how the three packages (Microsoft.DotNet.ILCompiler, Microsoft.NET.ILLink.Tasks, and runtime.\<rid\>.Microsoft.DotNet.ILCompiler) needed for NativeAOT are used.  It turns out that the runtime one, which is the large one, doesn't have any .props and .targets in it and is already downloaded as a PackageDownload.  So I think that package could be placed in the packs folder, and slight modifications to `ProcessFrameworkReferences` (probably mostly in the `AddToolPack` method) would allow it to be resolved from there.

The other two packages are way smaller, so I think including them as .nupkg files in the `library-packs` folder should work.  You still might want to change the package ID to something source-build specific if there are concerns that the Microsoft-built ones from NuGet.org could be used instead of the ones that ship with the source-built SDK in the library-packs folder.  We could also use the idea I had to allow these two packages to also be shipped in the packs folder if that was necessary.

As for fallback folders, I had forgotten about them, but was remembering the lessons we learned from them.  I believe fallback folders aren't supported (at least not well) anymore.  I believe we stopped using fallback folders around .NET Core 3, partly because we switched to framework references and package downloads, but also because fallback folders didn't work with features NuGet was adding such as package signing and hash validation.

**Comment by tmds** [comment]
I've started to implement the design proposed by @dsplaisted.

After patching the `ProcessFrameworkReferences.cs` so it looks for the source-built `runtime.<rid>.Microsoft.DotNet.ILCompiler` in `packs`, there is still another reference which triggers a nuget.org download for this package.

That reference comes from the `runtime.json` file that is in the source-built `Microsoft.DotNet.ILCompiler` package. It looks like:

```json
{
  "runtimes": {
    "fedora.39-x64": {
      "#import": [],
      "Microsoft.DotNet.ILCompiler": {
        "runtime.fedora.39-x64.Microsoft.DotNet.ILCompiler": "[9.0.0-preview.5.24272.2, )"
      }
    },
```
We need the `Microsoft.DotNet.ILCompiler` property to be empty, like:
```json
{
  "runtimes": {
    "fedora.39-x64": {
      "#import": [],
      "Microsoft.DotNet.ILCompiler": { }
    },
```

@dsplaisted @baronfel @ViktorHofer do you have a suggestion how/where to do this?

**Comment by dsplaisted** [comment]
Interesting.  I imagine that reference in the runtime.json file is there to support explicitly referencing the `Microsoft.DotNet.ILCompiler` package.  However, it seems that the PackageDownload added by the ProcessFrameworkReference task would be redundant.  @jkotas, do you know how this works?

I don't know where the `Microsoft.DotNet.ILCompiler` package is created, but for source build I think you could omit the runtime.json file entirely.

**Comment by jkotas** [comment]
Yes, explicit reference of Microsoft.DotNet.ILCompiler package is expected to add the RID-specific Microsoft.DotNet.ILCompiler package. runtime.json in Microsoft.DotNet.ILCompiler package on nuget.org has entries for all portable RIDs supported by native AOT to make that work.

**Comment by ViktorHofer** [comment]
The runtime.json feature is undocumented and got added in the .NET Core 1.x days to support including RID specific packages. The ILCompiler package is one of the remaining ones that use runtime.json and our plan is to eventually get rid off them all and replace them with alternative solutions (i.e. the Known* items in the SDK for anything in-built). https://github.com/dotnet/runtime/issues/49137 is also related.

Mentioning that as I think now would be a good time to change the ILCompiler package and make it use the common SDK infrastructure (framework references / known* items).

**Comment by dsplaisted** [comment]
> > explicit reference of Microsoft.DotNet.ILCompiler package is expected to add the RID-specific Microsoft.DotNet.ILCompiler package
> 
> I assume this means we can not just remove the `runtime.json` file.

I think the effect of removing the file entirely would be the same as removing the element you are suggesting.  The only thing that runtime.json file does is add the RID-specific package reference.  If you remove that element, then that should be the same as removing the entire file (except possibly for projects that are targeting other RuntimeIdentifiers).

I don't know if things will **work** if you remove the reference.  I'm not sure how the ILCompiler tasks locate the assets in the RID-specific ILCompiler package.  As a first step though I'd suggest trying out removing the runtime.json entirely and seeing what happens.

**Comment by jkotas** [comment]
> @jkotas are there cases where that won't work?

Is explicit reference of Microsoft.DotNet.ILCompiler package going to pick RID-specific Microsoft.DotNet.ILCompiler of the same version?

The explicit references of Microsoft.DotNet.ILCompiler are used e.g. here: https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/compiling.md#using-daily-builds

**Comment by tmds** [comment]
> Is explicit reference of Microsoft.DotNet.ILCompiler package going to pick RID-specific Microsoft.DotNet.ILCompiler of the same version?

I don't think so. We'd break this scenario by removing `runtime.json`.

**Comment by tmds** [comment]
How can we update/patch the `runtime.json` file so we eliminate the package reference only for the source-built rid?

**Comment by tmds** [comment]
@jkotas would it be ok if we'd remove `runtime.json` from `Microsoft.DotNet.ILCompiler` only when it gets built with source-build configuration?

**Comment by tmds** [comment]
> Is my understanding correct?

This is a good overview, and it matches with how I expect it to behave.

> Also, have you considered installing the native AOT compilers as separate distro package (e.g. dnf install dotnet-sdk-aot-8.0) to reduce the baseline size of .NET SDK? This package can transitively install [the native toolchain and other dependencies](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net7%2Clinux-ubuntu) so that users do not have to install them manually.

Yes, we'll put this in a separate package to reduce the default SDK size.

I like the idea of making this package install the native toolchain through dependencies. It will be up to the package maintainer to actually implement this.

We'll extend https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging with these guidelines.

**Comment by agocke** [comment]
This all sounds reasonable to me, but I want to make sure we have a long-term design in place that's specified and maintained. Do we have a doc that explains exactly what `library-packs` and `packs` directories are, how they interact with `PackageDownload`, how they are used as hooks for source build, and what semantics we have to maintain in the future?

This area has been difficult to understand in the past.

**Comment by tmds** [comment]
These are the PRs that implement https://github.com/dotnet/source-build/issues/1215#issuecomment-2116395577:
- https://github.com/dotnet/sdk/pull/41198
- https://github.com/dotnet/arcade/pull/14805
- https://github.com/dotnet/runtime/pull/102908

I verified that with these changes the resulting source-built SDK can be used to publish native-aot apps, and that it works as an SDK to build the vmr.

There is still some work to be done, esp on the sdk PR. I hope we can get these into the preview that the `main` branches are currently targeting.

Once these changes are in the vmr, we may look to do some optimizations that reduce the size of the ILCompiler pack.

**Comment by tmds** [comment]
@baronfel @dsplaisted @jkotas @MichaelSimons @ViktorHofer friendly reminder these PRs are still open, and this is something we'd really like to be in preview6.

cc @omajid

**Comment by MichaelSimons** [comment]
@tmds, @omajid, @ashnaga - Can we close this issue now?  I am thinking any remaining bugs found can be tracked with instance specific issues.

**Comment by tmds** [comment]
We still need to add tests, and update https://learn.microsoft.com/en-us/dotnet/core/distribution-packaging.

**Comment by MichaelSimons** [comment]
@tmds - Is testing the last piece here?  Is that still something you are planning to contribute?

**Comment by tmds** [comment]
Yes. I forgot about this. I'll look into it.
